spec.bs

```diff
@@ -6,57 +6,940 @@ Status: CG-DRAFT
+
+  A [=calling origin=]'s <dfn for="calling origin">remaining budget</dfn> is the [=entropy allowance=] minus any [=bit debits=] whose [=bit debit/timestamps=] are within the current [=budget epoch=].
+
+  {{WindowSharedStorage/selectURL()}}'s argument "`urls`" is its <dfn for=selectURL>input URL list</dfn>.
+
+  When an [=calling origin=] has insufficient [=calling origin/remaining budget=], {{WindowSharedStorage/selectURL()}} will return a {{FencedFrameConfig}} for the {{SharedStorageUrlWithMetadata/url}} in the {{SharedStorageUrlWithMetadata}} at the [=default index=] in its [=selectURL/input URL list=].
+
+  The <dfn>default index</dfn> for a call to {{WindowSharedStorage/selectURL()}} is implementation-defined in such a way that it is independent from the result of the associated {{SharedStorageSelectURLOperation}}'s "`run`" method.
+
+  <div class="example">
+    The [=default index=] could be defined to be 0.
+
+    In this case, whenever the {{SharedStorageSelectURLOperation}}'s "`run`" method encounters an error, or whenever there is insufficient [=calling origin/remaining budget=], the "`run`" method would return 0, and hence {{WindowSharedStorage/selectURL()}} would return the first {{SharedStorageUrlWithMetadata/url}} in its [=selectURL/input URL list=].
+  </div>
 
-  readonly attribute SharedStorageWorklet worklet;
-};
+  <div class="example">
+    The [=default index=] could be defined to be [=selectURL/input URL list=]'s [=list/size=] minus 1.
 
-[Exposed=(Window)]
-interface SharedStorageWorklet {
-  Promise<undefined> addModule(USVString moduleURL);
-};
+    In this case, whenever the {{SharedStorageSelectURLOperation}}'s "`run`" method encounters an error, or whenever there is insufficient [=calling origin/remaining budget=], {{WindowSharedStorage/selectURL()}} would return the last {{SharedStorageUrlWithMetadata/url}} in its [=selectURL/input URL list=].
+  </div>
 
-dictionary SharedStorageSetMethodOptions {
-  boolean ignoreIfPresent;
-};
+  <div algorithm>
+    To <dfn>determine remaining budget</dfn>, given a calling [=calling origin=] |origin|, run the following steps:
 
-dictionary SharedStorageRunOperationMethodOptions {
-  object data;
-};
+    1. If |origin| is [=opaque origin|opaque=], return undefined.
+    1. Let |maxBits| be the value of the [=user agent=]'s [=entropy allowance=].
+    1. If the [=user agent=]'s [=shared storage budget table=] does not [=map/contain=] |origin|, then return |maxBits|.
+    1. Otherwise, let |ledger| be [=user agent=]'s [=shared storage budget table=][|origin|].
+    1. Let |epochLength| be [=budget lifetime=].
+    1. Let |unsafeMoment| be the [=wall clock=]'s [=wall clock/unsafe current time=].
+    1. Let |moment| be the result of running [=coarsen time=] algorithm given |unsafeMoment| and [=wall clock=] as input.
+    1. Let |fromUnixEpochTime| be the [=duration from=] the [=Unix epoch=] to |moment|.
```

Comment:
Do we really need to convert into the epoch representation here?  Since this is all internal data it seems like we could just work with moments and durations?